darwin/CpuFreq.c

```diff
@@ -0,0 +1,176 @@
+        snprintf(buf, sizeof(buf), "IODeviceTree:/cpus/cpu%u", num_cpus);
+        io_registry_entry_t cpu_io_registry_entry = IORegistryEntryFromPath(kIOMainPortDefault, buf);
+        if (cpu_io_registry_entry == MACH_PORT_NULL) {
+            return -1;
+        }
+
+        CFDataRef cluster_type_ref = (CFDataRef) IORegistryEntryCreateCFProperty(cpu_io_registry_entry, CFSTR("cluster-type"), kCFAllocatorDefault, 0U);
+        if (cluster_type_ref == NULL) {
+            IOObjectRelease(cpu_io_registry_entry);
+            return -1;
+        }
+        if (CFDataGetLength(cluster_type_ref) != 2) {
+            CFRelease(cluster_type_ref);
+            IOObjectRelease(cpu_io_registry_entry);
+            return -1;
+        }
+        UniChar cluster_type_char;
+        CFDataGetBytes(cluster_type_ref, CFRangeMake(0, 2), (uint8_t*) &cluster_type_char);
+        CFRelease(cluster_type_ref);
+
+        uint32_t cluster_type = 0U;
+        switch (cluster_type_char) {
+            case L'E':
+                cluster_type = 0U;
+                break;
+            case L'P':
+                cluster_type = 1U;
+                break;
+            default:
+                /* Unknown cluster type */
+                IOObjectRelease(cpu_io_registry_entry);
+                return -1;
+        }
+
+        data->cluster_type_per_cpu[num_cpus] = cluster_type;
+
+        IOObjectRelease(cpu_io_registry_entry);
+    }
+
+    /*
+     * Determine frequencies for per-cluster-type performance states
+     * Frequencies for the "E" cluster type are stored in voltage-states1,
+     * frequencies for the "P" cluster type are stored in voltage-states5.
+     * This seems to be hardcoded.
+     */
+    const CFStringRef voltage_states_key_per_cluster[CPUFREQ_NUM_CLUSTER_TYPES] = {CFSTR("voltage-states1"), CFSTR("voltage-states5")};
+
+    io_registry_entry_t pmgr_registry_entry = IORegistryEntryFromPath(kIOMainPortDefault, "IODeviceTree:/arm-io/pmgr");
+    if (pmgr_registry_entry == MACH_PORT_NULL) {
+        return -1;
+    }
+    for (size_t i = 0U; i < CPUFREQ_NUM_CLUSTER_TYPES; i++) {
+        CFDataRef voltage_states_ref =
+            (CFDataRef) IORegistryEntryCreateCFProperty(pmgr_registry_entry, voltage_states_key_per_cluster[i], kCFAllocatorDefault, 0U);
+        if (voltage_states_ref == NULL) {
+            IOObjectRelease(pmgr_registry_entry);
+            return -1;
+        }
+
+        CpuFreqPowerStateFrequencies* cluster_frequencies = &data->cpu_frequencies_per_cluster_type[i];
+        cluster_frequencies->num_frequencies = CFDataGetLength(voltage_states_ref) / 8;
+        cluster_frequencies->frequencies = xCalloc(cluster_frequencies->num_frequencies, sizeof(double));
+        const uint8_t* voltage_states_data = CFDataGetBytePtr(voltage_states_ref);
+        for (size_t j = 0U; j < cluster_frequencies->num_frequencies; j++) {
+            uint32_t freq_value;
+            memcpy(&freq_value, voltage_states_data + j * 8, 4);
```

Comment:
Where is that 8 coming from?

Reply:
This is how these values are stored, in an unstructured byte array, every 8 bytes storing one frequency value. There is no documentation for this, it had to be determined based on trial and error (and matching it with the frequency values displayed by `powermetrics`).

Reply:
Would be nice to have some documentation of this somewhere. You could even do a packed structure with these values and reference those values that way.